### ISSUE TITLE
Root-implicit loop breaks out of optimization on a masked (fake) convergence

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#812]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

Split out of #792 / PR #811, which fixed the *reporting* half of this and deliberately left the control flow alone.

## The hole

In `optimize_gs_ad_root_implicit` (`ipeps_optimize_root_implicit.py`), a root solve that returns non-finite gradient entries is masked to zero:

```python
grads = jnp.where(jnp.isfinite(grads), grads, 0.0)
```

The loop itself already says what that means, in its own warning for the fully-masked case:

> every entry was non-finite, so the step is a no-op and **apparent convergence here is not convergence**

But nothing acts on that. A few lines later the masked gradient goes straight into the convergence test, and it satisfies it:

- `gs_conv_criterion='grad_norm'` — a fully masked gradient has L2 norm **exactly 0.0**, which is `< gs_grad_norm_tol` for any tolerance. Fires immediately.
- `gs_conv_criterion='dE'` (the current default) — a fully masked step is a no-op, so the next step re-evaluates identical params and sees `delta_energy == 0.0`. Fires one step later.

Either way the loop `break`s and returns as if the optimization had converged. Partial masking is milder but pushes the same direction: the masked entries contribute 0 to the norm, so `grad_norm` is deflated by exactly the components that failed.

## What #811 did and did not do

PR #811 added `return_history` to this path, which meant publishing a `converged` flag. It gates that flag on `nonfinite_grad_steps == 0`, so a contaminated run can no longer report `converged=True` — the flag stops contradicting the warning.

It does **not** change whether the loop breaks. That is pre-existing behaviour affecting every root-implicit caller, not just `return_history` ones, and changing it shifts returned energies and step counts, so it belongs on its own branch.

## Suggested fix

Skip the convergence test on any step whose gradient had non-finite entries, and reset `prev_energy = float("inf")` afterwards so the following step cannot false-converge on the resulting `delta_energy == 0.0`.

There is already precedent for exactly this in `ipeps_optimize.py`, in the grad-spike rollback path:

```python
# re-evaluates the same state and sees dE == 0 — which the "dE"
# convergence criterion would misread as convergence (the
# gradient is still large).  Reset prev_energy so the
# post-rollback step cannot false-converge.
prev_energy = float("inf")
continue
```

The root-implicit loop needs the same guard for the masking case.

## Related

Same family as #801 (measure the adjoint residual instead of trusting a flag that lies) and #772 (the `usable_rank` failure shape that produces these non-finite gradients in the first place). Refs #792.
